#include "bigheadpicturewindow.h"
#include "ui_bigheadpicturewindow.h"
#include <QFileDialog>
#include <QMessageBox>
#include <QStandardPaths>
#include <QCameraInfo>
#include <QCameraViewfinder>
#include <QCameraImageCapture>
#include <QDateTime>
#include <QPainter>
#include <QDir>
#include <QDebug>
#include <QResizeEvent>
#include <QElapsedTimer>
#include <QThread>

BigHeadPictureWindow::BigHeadPictureWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::BigHeadPictureWindow)
    , camera(nullptr)
    , viewfinder(nullptr)
    , imageCapture(nullptr)
    , cameraActive(false)
    , currentBgIndex(0)
{
    ui->setupUi(this);

    // åˆå§‹åŒ–UIçŠ¶æ€
    initUI();

    // åˆå§‹åŒ–ç›¸æœº
    initCamera();

    // åˆå§‹åŒ–èƒŒæ™¯å›¾ç‰‡
    initBackgrounds();

    // è¿æ¥ä¿¡å·æ§½
    initConnections();

    // æ›´æ–°UIçŠ¶æ€
    updateUI();
}

void BigHeadPictureWindow::initUI()
{
    // è®¾ç½®çª—å£å±æ€§
    setWindowTitle("å¤§å¤´ç…§ç›¸æœº");

    // åˆå§‹çŠ¶æ€
    ui->btnCapture->setEnabled(false);
    ui->btnPrevBg->setEnabled(false);
    ui->btnNextBg->setEnabled(true);

    // åˆ›å»ºç›¸æœºè§†å›¾å®¹å™¨
    viewfinder = new QCameraViewfinder(this);
    viewfinder->setMinimumSize(180, 180);
    viewfinder->setMaximumSize(180, 180);

    // è®¾ç½®åœ†å½¢é®ç½©
    viewfinder->setStyleSheet(
        "border-radius: 90px; "
        "background-color: transparent;"
        );

    // æ·»åŠ åˆ°ç›¸æœºå®¹å™¨
    ui->cameraContainerLayout->addWidget(viewfinder);
    viewfinder->setVisible(false);
}

void BigHeadPictureWindow::initCamera()
{
    // æ£€æŸ¥å¯ç”¨æ‘„åƒå¤´
    QList<QCameraInfo> cameras = QCameraInfo::availableCameras();
    if (cameras.isEmpty()) {
        qDebug() << "æœªæ£€æµ‹åˆ°æ‘„åƒå¤´";
        ui->btnToggleCamera->setEnabled(false);
        ui->btnToggleCamera->setText("æ— æ‘„åƒå¤´");
        ui->labelStatus->setText("æœªæ£€æµ‹åˆ°æ‘„åƒå¤´");
        return;
    }

    qDebug() << "æ‰¾åˆ°æ‘„åƒå¤´:" << cameras.size();

    // é€‰æ‹©ç¬¬ä¸€ä¸ªæ‘„åƒå¤´
    QCameraInfo selectedCamera = chooseCamera();

    // åˆ›å»ºç›¸æœºå¯¹è±¡
    camera = new QCamera(selectedCamera, this);
    imageCapture = new QCameraImageCapture(camera, this);

    // è®¾ç½®ç›¸æœºå‚æ•°ï¼ˆé™ä½åˆ†è¾¨ç‡å‡å°‘å†…å­˜ï¼‰
    QCameraViewfinderSettings settings;
    settings.setResolution(320, 240);  // ä½åˆ†è¾¨ç‡
    settings.setPixelFormat(QVideoFrame::Format_YUYV);
    settings.setMinimumFrameRate(10.0);
    settings.setMaximumFrameRate(20.0);

    camera->setViewfinderSettings(settings);
    camera->setViewfinder(viewfinder);

    // è¿æ¥ä¿¡å·
    connect(imageCapture, &QCameraImageCapture::imageCaptured,
            this, &BigHeadPictureWindow::onImageCaptured);
}
QCameraInfo BigHeadPictureWindow::chooseCamera(){
    const auto all = QCameraInfo::availableCameras();

    /* 1. ä¼˜å…ˆï¼šæè¿°é‡Œå¸¦ "USB" / "UVC" / "Web" çš„æ‘„åƒå¤´ */
    for (const QCameraInfo &i : all) {
        const QString desc = i.description().toLower();
        if (desc.contains("usb") || desc.contains("uvc") || desc.contains("web"))
            return i;
    }

    /* 2. å…¶æ¬¡ï¼šè®¾å¤‡åé‡Œå¸¦ video9 / video8 ... ç­‰ USB å£ */
    for (const QCameraInfo &i : all) {
        if (i.deviceName().contains("video9") ||
            i.deviceName().contains("video8") ||
            i.deviceName().contains("video7"))
            return i;
    }

    /* 3. æœ€åï¼šéšä¾¿ä¸€ä¸ªèƒ½ç”¨çš„ */
    return all.isEmpty() ? QCameraInfo() : all.first();
}
void BigHeadPictureWindow::initBackgrounds()
{
    // æ¸…ç©ºèƒŒæ™¯åˆ—è¡¨
    backgrounds.clear();
    bgNames.clear();

    // æ·»åŠ ç®€å•èƒŒæ™¯ï¼ˆçº¯è‰²ï¼‰
    // èƒŒæ™¯1ï¼šç™½è‰²
    backgrounds.append(QPixmap(400, 400));
    backgrounds.last().fill(Qt::white);
    bgNames.append("çº¯ç™½èƒŒæ™¯");

    // èƒŒæ™¯2ï¼šæµ…ç°è‰²
    backgrounds.append(QPixmap(400, 400));
    backgrounds.last().fill(Qt::lightGray);
    bgNames.append("æµ…ç°èƒŒæ™¯");

    // èƒŒæ™¯3ï¼šæµ…è“è‰²
    backgrounds.append(QPixmap(400, 400));
    backgrounds.last().fill(QColor(173, 216, 230));
    bgNames.append("æµ…è“èƒŒæ™¯");

    // èƒŒæ™¯4ï¼šæµ…ç²‰è‰²
    backgrounds.append(QPixmap(400, 400));
    backgrounds.last().fill(QColor(255, 182, 193));
    bgNames.append("æµ…ç²‰èƒŒæ™¯");

    // èƒŒæ™¯5ï¼šæµ…ç»¿è‰²
    backgrounds.append(QPixmap(400, 400));
    backgrounds.last().fill(QColor(152, 251, 152));
    bgNames.append("æµ…ç»¿èƒŒæ™¯");

    // èƒŒæ™¯6ï¼šæ¸å˜èƒŒæ™¯
    QPixmap gradientBg(400, 400);
    QPainter painter(&gradientBg);
    QLinearGradient gradient(0, 0, 400, 400);
    gradient.setColorAt(0, Qt::cyan);
    gradient.setColorAt(1, Qt::blue);
    painter.fillRect(gradientBg.rect(), gradient);
    painter.end();
    backgrounds.append(gradientBg);
    bgNames.append("æ¸å˜èƒŒæ™¯");

    // æ›´æ–°èƒŒæ™¯æ˜¾ç¤º
    // updateBackgroundDisplay();
}

void BigHeadPictureWindow::initConnections()
{
    // ç›¸æœºæ§åˆ¶
    connect(ui->btnToggleCamera, &QPushButton::clicked,
            this, &BigHeadPictureWindow::onBtnToggleCameraClicked);
    connect(ui->btnCapture, &QPushButton::clicked,
            this, &BigHeadPictureWindow::onBtnCaptureClicked);

    // èƒŒæ™¯æ§åˆ¶
    connect(ui->btnPrevBg, &QPushButton::clicked,
            this, &BigHeadPictureWindow::onBtnPrevBgClicked);
    connect(ui->btnNextBg, &QPushButton::clicked,
            this, &BigHeadPictureWindow::onBtnNextBgClicked);
}

void BigHeadPictureWindow::updateUI()
{
    // æ›´æ–°èƒŒæ™¯åç§°æ ‡ç­¾
    if (currentBgIndex >= 0 && currentBgIndex < bgNames.size()) {
        ui->labelBgName->setText(bgNames[currentBgIndex]);
    }

    // æ›´æ–°æŒ‰é’®çŠ¶æ€
    ui->btnPrevBg->setEnabled(currentBgIndex > 0);
    ui->btnNextBg->setEnabled(currentBgIndex < backgrounds.size() - 1);

    // æ›´æ–°çŠ¶æ€æ 
    ui->labelStatus->setText(QString("å½“å‰èƒŒæ™¯: %1/%2")
                                 .arg(currentBgIndex + 1)
                                 .arg(backgrounds.size()));
}

void BigHeadPictureWindow::updateBackgroundDisplay()
{
    if (currentBgIndex >= 0 && currentBgIndex < backgrounds.size()) {
        QPixmap bg = backgrounds[currentBgIndex];

        // ç¼©æ”¾èƒŒæ™¯ä»¥é€‚åº”æ˜¾ç¤ºåŒºåŸŸ
        QSize labelSize = ui->labelBackground->size();
        QPixmap scaledBg = bg.scaled(labelSize, Qt::KeepAspectRatioByExpanding, Qt::SmoothTransformation);

        ui->labelBackground->setPixmap(scaledBg);
    }
}

void BigHeadPictureWindow::resizeEvent(QResizeEvent *event)
{
    QMainWindow::resizeEvent(event);
    updateBackgroundDisplay();
}

void BigHeadPictureWindow::onBtnToggleCameraClicked()
{
    if (!camera) {
        QMessageBox::warning(this, "é”™è¯¯", "æ‘„åƒå¤´åˆå§‹åŒ–å¤±è´¥");
        return;
    }

    if (!cameraActive) {
        // å¼€å¯ç›¸æœº
        try {
            camera->start();

            // ç­‰å¾…ç›¸æœºå¯åŠ¨
            QElapsedTimer timer;
            timer.start();
            while (camera->state() != QCamera::ActiveState && timer.elapsed() < 3000) {
                QCoreApplication::processEvents();
                QThread::msleep(50);
            }

            if (camera->state() == QCamera::ActiveState) {
                viewfinder->setVisible(true);
                ui->btnToggleCamera->setText("ğŸ“· å…³é—­ç›¸æœº");
                ui->btnCapture->setEnabled(true);
                cameraActive = true;
                ui->labelStatus->setText("ç›¸æœºè¿è¡Œä¸­ - " + bgNames[currentBgIndex]);
            } else {
                QMessageBox::warning(this, "é”™è¯¯", "ç›¸æœºå¯åŠ¨å¤±è´¥");
                camera->stop();
            }
        } catch (...) {
            QMessageBox::critical(this, "é”™è¯¯", "ç›¸æœºå¯åŠ¨å¼‚å¸¸");
        }
    } else {
        // å…³é—­ç›¸æœº
        camera->stop();
        viewfinder->setVisible(false);
        ui->btnToggleCamera->setText("ğŸ“· å¼€å¯ç›¸æœº");
        ui->btnCapture->setEnabled(false);
        cameraActive = false;
        ui->labelStatus->setText("ç›¸æœºå·²åœæ­¢ - " + bgNames[currentBgIndex]);
    }
}

void BigHeadPictureWindow::onBtnCaptureClicked()
{
    if (!cameraActive || !imageCapture) {
        QMessageBox::warning(this, "é”™è¯¯", "è¯·å…ˆå¼€å¯ç›¸æœº");
        return;
    }

    ui->btnCapture->setEnabled(false);
    ui->labelStatus->setText("æ­£åœ¨æ‹ç…§...");

    // æ•è·å›¾åƒ
    imageCapture->capture();
}

void BigHeadPictureWindow::onImageCaptured(int id, const QImage &preview)
{
    Q_UNUSED(id);

    // åˆæˆå¤§å¤´ç…§
    QPixmap result = combineHeadPicture(preview);

    // ä¿å­˜åˆ°æ–‡ä»¶
    saveHeadPicture(result);

    // æ¢å¤æŒ‰é’®çŠ¶æ€
    ui->btnCapture->setEnabled(true);
}

QPixmap BigHeadPictureWindow::combineHeadPicture(const QImage &cameraImage)
{
    // åˆ›å»ºç»“æœå›¾åƒï¼ˆä¸èƒŒæ™¯ç›¸åŒå¤§å°ï¼‰
    QPixmap result(400, 400);

    // ç¬¬ä¸€æ­¥ï¼šç»˜åˆ¶èƒŒæ™¯
    QPainter painter(&result);
    painter.setRenderHint(QPainter::Antialiasing);

    if (currentBgIndex >= 0 && currentBgIndex < backgrounds.size()) {
        painter.drawPixmap(0, 0, backgrounds[currentBgIndex]);
    } else {
        painter.fillRect(result.rect(), Qt::white);
    }

    // ç¬¬äºŒæ­¥ï¼šç»˜åˆ¶ç›¸æœºå›¾åƒï¼ˆåœ†å½¢å¤´åƒï¼‰
    if (!cameraImage.isNull()) {
        // ç¼©æ”¾ç›¸æœºå›¾åƒä»¥é€‚åº”åœ†å½¢åŒºåŸŸ
        QImage scaledCamera = cameraImage.scaled(200, 200, Qt::KeepAspectRatioByExpanding, Qt::SmoothTransformation);

        // åˆ›å»ºåœ†å½¢è’™ç‰ˆ
        QPixmap mask(200, 200);
        mask.fill(Qt::transparent);

        QPainter maskPainter(&mask);
        maskPainter.setRenderHint(QPainter::Antialiasing);
        maskPainter.setBrush(Qt::black);
        maskPainter.setPen(Qt::NoPen);
        maskPainter.drawEllipse(0, 0, 200, 200);
        maskPainter.end();

        // åº”ç”¨åœ†å½¢è’™ç‰ˆ
        QImage circularImage(200, 200, QImage::Format_ARGB32);
        circularImage.fill(Qt::transparent);

        for (int y = 0; y < 200; ++y) {
            for (int x = 0; x < 200; ++x) {
                QRgb maskPixel = mask.toImage().pixel(x, y);
                if (qAlpha(maskPixel) > 0) {
                    int srcX = (x * scaledCamera.width()) / 200;
                    int srcY = (y * scaledCamera.height()) / 200;

                    if (srcX < scaledCamera.width() && srcY < scaledCamera.height()) {
                        circularImage.setPixel(x, y, scaledCamera.pixel(srcX, srcY));
                    }
                }
            }
        }

        // åœ¨ä¸­å¿ƒä½ç½®ç»˜åˆ¶åœ†å½¢å¤´åƒ
        int xPos = (400 - 200) / 2;
        int yPos = (400 - 200) / 2;
        painter.drawImage(xPos, yPos, circularImage);

        // æ·»åŠ ç™½è‰²è¾¹æ¡†
        painter.setPen(QPen(Qt::white, 3));
        painter.setBrush(Qt::NoBrush);
        painter.drawEllipse(xPos, yPos, 200, 200);
    }

    painter.end();

    return result;
}

void BigHeadPictureWindow::saveHeadPicture(const QPixmap &picture)
{
    // åˆ›å»ºä¿å­˜ç›®å½•
    QString picturesDir = QStandardPaths::writableLocation(QStandardPaths::PicturesLocation);
    QString saveDir = picturesDir + "/BigHeadPictures";
    QDir().mkpath(saveDir);

    // ç”Ÿæˆæ–‡ä»¶å
    QString timestamp = QDateTime::currentDateTime().toString("yyyyMMdd_hhmmss");
    QString fileName = saveDir + "/bighead_" + timestamp + ".jpg";

    // ä¿å­˜å›¾åƒï¼ˆé™ä½è´¨é‡ä»¥å‡å°‘æ–‡ä»¶å¤§å°ï¼‰
    if (picture.save(fileName, "JPG", 80)) {
        ui->labelStatus->setText("ä¿å­˜æˆåŠŸ: " + bgNames[currentBgIndex]);
        QMessageBox::information(this, "ä¿å­˜æˆåŠŸ",
                                 QString("å¤§å¤´ç…§å·²ä¿å­˜\nèƒŒæ™¯: %1").arg(bgNames[currentBgIndex]));
    } else {
        ui->labelStatus->setText("ä¿å­˜å¤±è´¥");
        QMessageBox::warning(this, "ä¿å­˜å¤±è´¥", "æ— æ³•ä¿å­˜å›¾ç‰‡");
    }
}

void BigHeadPictureWindow::onBtnPrevBgClicked()
{
    if (currentBgIndex > 0) {
        currentBgIndex--;
        //updateBackgroundDisplay();
        updateUI();
    }
}

void BigHeadPictureWindow::onBtnNextBgClicked()
{
    if (currentBgIndex < backgrounds.size() - 1) {
        currentBgIndex++;
        //updateBackgroundDisplay();
        updateUI();
    }
}

BigHeadPictureWindow::~BigHeadPictureWindow()
{
    // åœæ­¢ç›¸æœº
    if (camera && cameraActive) {
        camera->stop();
    }

    delete ui;
}
